### PR TITLE
[NPM-3531] Add TCP state tracking to ebpf-less

### PR DESCRIPTION
**Pull request by pimlu**
### What does this PR do?
This PR separates the TCPProcessor into its own file and adds logic tracking state according to the TCP RFC's state machine.

### Motivation
We need this to keep track of tcp connections opening/closing, retransmits, etc

### Describe how to test/QA your changes

To just test the TCP state machine without running the tracer:
```
go test -tags=linux,linux_bpf,npm,process,test ./pkg/network/tracer/connection/ebpfless
```

To run the ebpfless tracer end to end on the test suite:
```
 DD_REMOTE_CONFIGURATION_ENABLED=false TEST_EBPFLESS_OVERRIDE=true sudo -E go test -tags=linux,linux_bpf,npm,process,test ./pkg/network/tracer -v --run TestTracerSuite/eBPFless
```
### Possible Drawbacks / Trade-offs

Currently TCP simultaneous open is not quite tracked correctly (~~`RecvBytes` is one off,~~ and it decides the connection is established 1 packet too early) but it is good enough for now.

Might not be possible to pass TestSelfConnect because each packet is captured twice on the send/receiving end respectively. We could filter out this case based on `srcIP==dstIP && srcPort == dstPort`

Right now, doesn't implement pre-existing connections (I think it needs to populate the connections from netlink)

Connection durations are not tracked yet (this is probably simple, just not sure how yet).

### Additional Notes

Future things to implement:
* RTT
* TCP retransmit detection (I put TODOs in the places where retransmits occur)

### CODE REVIEW

File: pkg/network/tracer/tracer_test.go [hunk 1]
```diff
@@ -390,14 +390,21 @@ func (s *TracerSuite) TestTCPConnsReported() {
...
-	// Test
-	connections := getConnections(t, tr)
-	// Server-side
-	_, ok := findConnection(c.RemoteAddr(), c.LocalAddr(), connections)
-	require.True(t, ok)
-	// Client-side
-	_, ok = findConnection(c.LocalAddr(), c.RemoteAddr(), connections)
-	require.True(t, ok)
+	if tr.config.EnableEbpfless {
+		time.Sleep(time.Millisecond * 500)
```
Comment (hmahmood) [inline pkg/network/tracer/tracer_test.go:394]: Do we need this given the `require.Eventually` below?

File: pkg/network/tracer/connection/ebpfless_tracer.go [hunk 2]
```diff
@@ -199,9 +200,15 @@ func (t *ebpfLessTracer) processConnection(
 	var err error
 	switch conn.Type {
 	case network.UDP:
+		if (ip4 != nil && !t.config.CollectUDPv4Conns) || (ip6 != nil && !t.config.CollectUDPv6Conns) {
```
Comment (hmahmood) [inline pkg/network/tracer/connection/ebpfless_tracer.go:203]: Should we also just skip if `ip4/6 == nil`? Ditto for tcp

File: pkg/network/tracer/connection/ebpfless/tcp_processor.go [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+				if isFinAck {
+					ackDiff--
+				}
+				conn.Monotonic.RecvBytes += uint64(ackDiff)
+				st.lastAck = tcp.Ack
+			}
+		}
+	case unix.PACKET_HOST:
+		conn.Monotonic.RecvPackets++
+		if !st.hasReceivedPacket {
```
Comment (hmahmood) [inline pkg/network/tracer/connection/ebpfless/tcp_processor.go:109]: You could also just use `conn.Monotonic.RecvPackets` here.

File: pkg/network/tracer/connection/ebpfless/tcp_processor.go [hunk 4]
```diff
@@ -0,0 +1,338 @@
...
+		packetInfo := debugPacketInfo(pktType, tcp, payloadLen, st)
+		return "tcp processor: " + packetInfo
+	})
+
+	// skip invalid packets we don't recognize.
+	synFinCombo := tcp.SYN && tcp.FIN
+	noFlagsCombo := !tcp.SYN && !tcp.FIN && !tcp.ACK && !tcp.RST
+	if synFinCombo || noFlagsCombo {
+		// TODO remove this warning since probably these occur in the wild
+		log.Warnf("invalid flags combo: SYN=%t FIN=%t ACK=%t RST=%t", tcp.SYN, tcp.FIN, tcp.ACK, tcp.RST)
```
Comment (hmahmood) [inline pkg/network/tracer/connection/ebpfless/tcp_processor.go:127]: You could also add some telemetry here.

File: pkg/network/tracer/tracer_test.go (hunk 1 above)
Comment (pimlu) [reply pkg/network/tracer/tracer_test.go:394]: Ah whoops, I meant for the require.Eventually to replace this - thanks

File: pkg/network/tracer/connection/ebpfless/tcp_processor.go [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+		ackOutdated := !st.hasRemoteAck || isSeqBefore(st.lastRemoteAck, tcp.Ack)
+		if tcp.ACK && ackOutdated {
+			st.hasRemoteAck = true
+			st.lastRemoteAck = tcp.Ack
+		}
+	}
+}
+
+func (t *TCPProcessor) updateFinFlag(conn *network.ConnectionStats, st *connectionState, pktType uint8, tcp *layers.TCP, payloadLen uint16) {
+	payloadSeq := tcp.Seq + uint32(payloadLen)
```
Comment (hmahmood) [inline pkg/network/tracer/connection/ebpfless/tcp_processor.go:126]: Do we want to return early if the connection is already marked as closed?

File: pkg/network/tracer/connection/ebpfless_tracer.go (hunk 2 above)
Comment (pimlu) [reply pkg/network/tracer/connection/ebpfless_tracer.go]: Done in the previous UDP PR

File: pkg/network/tracer/connection/ebpfless/tcp_processor.go (hunk 5 above)
Comment (pimlu) [reply pkg/network/tracer/connection/ebpfless/tcp_processor.go:126]: Oh that is a good idea

File: pkg/network/tracer/connection/ebpfless/tcp_processor.go (hunk 4 above)
Comment (pimlu) [reply pkg/network/tracer/connection/ebpfless/tcp_processor.go]: Switched this to telemetry

File: pkg/network/tracer/connection/ebpfless/tcp_processor.go (hunk 5 above)
Comment (pimlu) [reply pkg/network/tracer/connection/ebpfless/tcp_processor.go:126]: Hm actually, I might need to keep running through this logic post-close to handle FIN retransmits in the future

File: pkg/network/tracer/connection/ebpfless/tcp_processor.go (hunk 3 above)
Comment (pimlu) [reply pkg/network/tracer/connection/ebpfless/tcp_processor.go]: I added `hasReceivedPacket` under the impression that `RecvPackets` would remain nonzero/keep climbing if a connection tuple is reopened a second/third time in between network checks. Network checks drain out the `ConnectionStats` structure right?

**Review by hmahmood — APPROVED** [review]

**Pull request merged**